contracts/modules/capital/SwapOperator.sol

```diff
@@ -86,129 +78,254 @@ contract SwapOperator {
+  /// Skips validation for WETH since it does not have set swapDetails
+  function validateSellTokenBalance(
+    IPool pool,
+    GPv2Order.Data calldata order,
+    SwapDetails memory sellSwapDetails,
+    SwapOperationType swapOperationType,
+    uint totalOutAmount
+  ) internal view {
+    uint sellTokenBalance = order.sellToken.balanceOf(address(pool));
 
-    // Validate basic CoW params
-    validateBasicCowParams(order);
+    // validate ETH balance is within ETH reserves after the swap
+    if (swapOperationType == SwapOperationType.EthToAsset) {
+      uint ethPostSwap = address(pool).balance - totalOutAmount;
+      if (ethPostSwap < minPoolEth) {
+        revert EthReserveBelowMin(ethPostSwap, minPoolEth);
+      }
+      // skip sellSwapDetails validation for ETH/WETH since it does not have set swapDetails
+      return;
+    }
 
-    IPool pool = _pool();
-    IPriceFeedOracle priceFeedOracle = pool.priceFeedOracle();
-    uint totalOutAmount = order.sellAmount + order.feeAmount;
+    if (sellTokenBalance <= sellSwapDetails.maxAmount) {
+      revert InvalidBalance(sellTokenBalance, sellSwapDetails.maxAmount, 'max');
+    }
+    // NOTE: the totalOutAmount (i.e. sellAmount + fee) is used to get postSellTokenSwapBalance
+    uint postSellTokenSwapBalance = sellTokenBalance - totalOutAmount; 
+    if (postSellTokenSwapBalance < sellSwapDetails.minAmount) {
+      revert InvalidPostSwapBalance(postSellTokenSwapBalance, sellSwapDetails.minAmount, 'min');
+    }      
+  }
 
-    if (isSellingEth(order)) {
-      // Validate min/max setup for buyToken
-      SwapDetails memory swapDetails = pool.getAssetSwapDetails(address(order.buyToken));
-      require(swapDetails.minAmount != 0 || swapDetails.maxAmount != 0, "SwapOp: buyToken is not enabled");
-      uint buyTokenBalance = order.buyToken.balanceOf(address(pool));
-      require(buyTokenBalance < swapDetails.minAmount, "SwapOp: can only buy asset when < minAmount");
-      require(buyTokenBalance + order.buyAmount <= swapDetails.maxAmount, "SwapOp: swap brings buyToken above max");
+  /// @dev Validates two conditions:
+  /// 1. The current buyToken balance is less than buySwapDetails.minAmount
+  /// 2. The post-swap buyToken balance is less than or equal to buySwapDetails.maxAmount
+  /// Skip validation for WETH since it does not have set swapDetails
+  function validateBuyTokenBalance(
+    IPool pool,
+    GPv2Order.Data calldata order,
+    SwapDetails memory buySwapDetails
+  ) internal view {
+    uint buyTokenBalance = order.buyToken.balanceOf(address(pool));
+    
+    // skip validation for WETH since it does not have set swapDetails
+    if (address(order.buyToken) == address(weth)) {
+      return;
+    }
 
-      validateSwapFrequency(swapDetails);
+    if (buyTokenBalance >= buySwapDetails.minAmount) {
+      revert InvalidBalance(buyTokenBalance, buySwapDetails.minAmount, 'min');
+    }
```

Comment:
We're essentially using the `minAmount` value here as a max limit  (i.e. buyTokenBalance cannot be >= the minAmount)

original code:
```solidity
require(buyTokenBalance < swapDetails.minAmount, "SwapOp: can only buy asset when < minAmount");
```

So it might be more appropriate to set the limit type as 'max':

```solidity
revert InvalidBalance(buyTokenBalance, buySwapDetails.minAmount, 'max');
```

@shark0der thoughts?

Reply:
thinking about this further I'm tempted to drop the 3rd param altogether (limit type).

Because if the given `value` is less than the `limitAmount`, we can deduce its a minimum threshold and vice versa for maximum threshold.

The 3rd parameter can be a source of confusion if the values are used weirdly like the example above.